models/accm.py

```diff
@@ -30,14 +31,24 @@ def forward(self, x):
 
 
 class AccmBlock(nn.Module):
-    def __init__(self, input_layer, h_channels, number_of_rsacm):
+    def __init__(self, input_layer, h_channels, number_of_rsacm, args, rsacm_cost):
         super().__init__()
         self.h_channels = h_channels
         self.number_of_rsacm = number_of_rsacm
+        self.args = args
         self.rsacm_list = nn.ModuleList(
             [(RsacmBlock(input_layer, self.h_channels)) for i in range(0, self.number_of_rsacm)])
+        self.rsacm_cost = rsacm_cost
+
+    def initialize_gating_network(self, activation, args):
+        self.gating_network =  Gating_Network(activation, args)
+        self.optimizer_for_gating_network = torch.optim.Adam(self.gating_network.parameters(), args.lr, weight_decay=0)
 
     def forward(self, x):
+        if self.args.train_gating_networks:
+            gating_network_output =  self.gating_network(x)
+            gating_network_output_index = torch.argmax(gating_network_output, dim=1)

```

Comment:
Argmax is not differentiable, and we want the gradient to flow to the gating network (to all of them, actually). Softmax is actually the differentiable approximation to argmax we want (and, in particular, gumbel-softmax). We can:
1. optionally leave the argmax for inference (so that we are honest in the fact that our model really saves FLOPs),
2. use the gating network output as-is (the output after gumbel-softmax).
I think that point 2 means that the gating network should return $N$ outputs for each example, instead of an index.